### PR TITLE
lobby camera

### PR DESCRIPTION
**Pull request by jjpark-kb**
## About The Pull Request
adds a lobby camera. 
Original code is by Falaskian from Toolbox Station, but this has been modified.
https://gyazo.com/dc4497e78297b0138b1ea4b4aa5fe15e

## How This Contributes To The Skyrat Roleplay Experience
a nice rolling camera intro before the round

## Changelog

:cl:
add: added the lobby camera
/:cl:

### CODE REVIEW

**Comment by Zergspower** [comment]
The menu disappears when the cam shifts

File: code/__DEFINES/layers.dm [hunk 1]
```diff
@@ -183,9 +183,12 @@
...
 #define ADMIN_POPUP_LAYER 1
 
-
-///Plane of the "splash" icon used that shows on the lobby screen. Nothing should ever be above this.
-#define SPLASHSCREEN_PLANE 9999
+//SKYRAT EDIT: Lobby Cam
+///Plane of the "splash" icon used that shows on the lobby screen. Nothing should ever be above this. (Except the logo)
+#define SPLASHSCREEN_PLANE 9997 //Previous value: 9999
+#define BLACK_FADE 9998
+#define SKYRAT_LOGO 9999
```
Comment (jjpark-kb) [inline code/__DEFINES/layers.dm]:
```suggestion
#define BLACK_FADE 9997
#define SPLASHSCREEN_PLANE 9998 //Previous value: 9999
#define SKYRAT_LOGO 9999
```

File: modular_skyrat/modules/lobby_cam/code/lobby_cam.dm [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+/obj/new_player_cam/proc/fade_logo()
+	for(var/mob/checking_mob in GLOB.new_player_list)
+		var/atom/movable/screen/skyrat_logo/logo_screen = locate() in checking_mob.client.screen
+		if(logo_screen?.alpha == 255)
+			animate(logo_screen, alpha = 0, time = 5 SECONDS)
+
+/obj/new_player_cam/New()
+	fading_screen = new()
+	spawn(0)
```
Comment (Iamgoofball) [inline modular_skyrat/modules/lobby_cam/code/lobby_cam.dm]: how old is this code?

**Comment by desertmote** [comment]
Some issues: It switches really quickly, and the black fade **also fades the menu options**.

**Comment by jjpark-kb** [comment]
the menu being covered has been resolved. I tried lowering the speed and it makes it look far more jittery.

**Comment by desertmote** [comment]
is there another way the duration of the camera panning can be extended? good job on the menu thing btw

edit: tweak `path_target` and `moving_speed` maybe

**Comment by SpaceVampire** [comment]
It feels really jittery, to the point that I feel actually motion sick from looking at it. Really don't like trying to focus on OOC chat with something moving all around on the left side of my screen. Could you at least add some option for people to disable it, if they don't want it? I'm kind of sad about how this is going to be the death of cool custom title screens, but I guess that part is whatever.

**Comment by YakumoChen** [comment]
The map isn't really designed to be viewed pre-round. You can see all the random item spawn markers and such. It doesn't look good.

**Comment by desertmote** [comment]
I noticed you tweaked the numbers for how the camera moves! It feels much nicer now.

I have something else to nitpick about: Is there a way to make the camera initiate later? Particularly so its loaded after the part where all lights in the map turn on? This is a Skyrat thing and has always been hugely laggy/ugly to witness, it would be nice if the lobby-camera started just after all lights were turned on.

File: modular_skyrat/modules/lobby_cam/code/lobby_eye_ss.dm [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+/datum/controller/subsystem/lobby_eye/Initialize(start_timeofday)
+	. = ..()
+	linked_camera = new()
+	fading_screen = new()
+	fire()
+
+//fades out the skyrat logo in the bottom left
+/datum/controller/subsystem/lobby_eye/proc/fade_logo()
+	for(var/mob/checking_mob in GLOB.new_player_list)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/lobby_cam/code/lobby_eye_ss.dm]:
```suggestion
	for(var/mob/checking_mob as anything in GLOB.new_player_list)
```

File: modular_skyrat/modules/lobby_cam/code/lobby_eye_ss.dm [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		var/atom/movable/screen/skyrat_logo/logo_screen = locate() in checking_mob.client.screen
+		if(logo_screen?.alpha == 255)
+			animate(logo_screen, alpha = 0, time = 5 SECONDS)
+
+//gets everyone on the camera
+/datum/controller/subsystem/lobby_eye/proc/lock_eyes()
+	for(var/mob/checking_mob in GLOB.new_player_list)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/lobby_cam/code/lobby_eye_ss.dm]: Pretty sure this list shouldn't contain anything else than that.
```suggestion
	for(var/mob/checking_mob as anything in GLOB.new_player_list)
```

File: modular_skyrat/modules/lobby_cam/code/lobby_eye_ss.dm [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+		if(!checking_mob.client)
+			continue
+		if(!checking_mob.client.prefs.read_preference(/datum/preference/toggle/lobby_cam))
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/lobby_cam/code/lobby_eye_ss.dm]: Clients are finnicky, let's avoid runtimes.
```suggestion
		if(!checking_mob.client?.prefs.read_preference(/datum/preference/toggle/lobby_cam))
```

File: modular_skyrat/modules/lobby_cam/code/lobby_eye_ss.dm [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+			continue
+		if(SSticker.current_state < GAME_STATE_PREGAME)
+			continue
+		if(!(fading_screen in checking_mob.client.screen))
+			checking_mob.client.screen += fading_screen
+		if(checking_mob.client.eye != linked_camera)
+			checking_mob.client.eye = linked_camera
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/lobby_cam/code/lobby_eye_ss.dm]:
```suggestion
		if(!(fading_screen in checking_mob.client?.screen))
			checking_mob.client?.screen += fading_screen
		if(checking_mob.client?.eye != linked_camera)
			checking_mob.client?.eye = linked_camera
```

File: modular_skyrat/modules/lobby_cam/code/lobby_eye_ss.dm [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+//gets everyone off of the camera
+/datum/controller/subsystem/lobby_eye/proc/unlock_eyes()
+	for(var/mob/checking_mob in GLOB.new_player_list)
+		if(!checking_mob.client)
+			continue
+		if(fading_screen && (fading_screen in checking_mob.client.screen))
+			checking_mob.client.screen -= fading_screen
+		if(checking_mob.client.eye != linked_camera)
+			continue
+		checking_mob.client.eye = checking_mob.client.mob
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/lobby_cam/code/lobby_eye_ss.dm]:
```suggestion
		if(fading_screen && (fading_screen in checking_mob.client?.screen))
			checking_mob.client?.screen -= fading_screen
		if(checking_mob.client?.eye != linked_camera)
			continue
		checking_mob.client?.eye = checking_mob.client?.mob
```

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]
Aside from that minor nitpick, that code seems rather solid.

**Pull request merged**